miniapps/meshing/minimal-surface.cpp

```diff
@@ -0,0 +1,1302 @@
+      bool Converged(const double rnorm) { return rnorm < NRM; }
+
+      bool ParAXeqB()
+      {
+         b = 0.0;
+         Vector X, B;
+         a.FormLinearSystem(S.bc, x, b, A, X, B);
+         if (!opt.pa) { M = new GSSmoother((SparseMatrix&)(*A)); }
+         if (M) { cg.SetPreconditioner(*M); }
+         cg.SetOperator(*A);
+         cg.Mult(B, X);
+         a.RecoverFEMSolution(X, b, x);
+         const bool by_vdim = opt.by_vdim;
+         GridFunction *nodes = by_vdim ? &x0 : S.fes->GetMesh()->GetNodes();
+         x.HostReadWrite();
+         nodes->HostRead();
+         double rnorm = nodes->DistanceTo(x) / nodes->Norml2();
+         if (!opt.id) { mfem::out << "rnorm = " << rnorm << std::endl; }
+         const double lambda = opt.lambda;
+         if (by_vdim)
+         {
+            MFEM_VERIFY(!opt.radial,"'VDim solver can't use radial option!");
+            return Converged(rnorm);
+         }
+         if (opt.radial)
+         {
+            GridFunction delta(S.fes);
+            subtract(x, *nodes, delta); // Δ = x - nodes
+            // position and Δ vectors at point i
+            Vector ni(SDIM), di(SDIM);
+            for (int i = 0; i < delta.Size()/SDIM; i++)
+            {
+               // extract local vectors
+               const int ndof = S.fes->GetNDofs();
+               for (int d = 0; d < SDIM; d++)
+               {
+                  ni(d) = (*nodes)(d*ndof + i);
+                  di(d) = delta(d*ndof + i);
+               }
+               // project the delta vector in radial direction
+               const double ndotd = (ni*di) / (ni*ni);
+               di.Set(ndotd,ni);
+               // set global vectors
+               for (int d = 0; d < SDIM; d++) { delta(d*ndof + i) = di(d); }
+            }
+            add(*nodes, delta, *nodes);
+         }
+         // x = λ*nodes + (1-λ)*x
+         add(lambda, *nodes, (1.0-lambda), x, x);
+         return Converged(rnorm);
+      }
+
+      void Amr()
+      {
+         MFEM_VERIFY(opt.amr_threshold >= 0.0 && opt.amr_threshold <= 1.0, "");
+         Mesh *mesh = S.mesh;
+         Array<Refinement> amr;
+         const int NE = mesh->GetNE();
+         DenseMatrix Jadjt, Jadj(DIM, SDIM);
+         for (int e = 0; e < NE; e++)
+         {
+            double minW = +NL_DMAX;
+            double maxW = -NL_DMAX;
+            ElementTransformation *eTr = mesh->GetElementTransformation(e);
+            const Geometry::Type &type = mesh->GetElement(e)->GetGeometryType();
+            const IntegrationRule *ir = &IntRules.Get(type, opt.order);
+            const int NQ = ir->GetNPoints();
+            for (int q = 0; q < NQ; q++)
+            {
+               eTr->SetIntPoint(&ir->IntPoint(q));
+               const DenseMatrix &J = eTr->Jacobian();
+               CalcAdjugate(J, Jadj);
+               Jadjt = Jadj;
+               Jadjt.Transpose();
+               const double w = Jadjt.Weight();
+               minW = fmin(minW, w);
```

Comment:
A matter of taste, but I always prefer to explicitly use the `std::` namespace (for `fmin` in this case) and explicitly use the proper include inside the compilation unit

Reply:
sure, `done`